src/camera_autoexposure.cpp

```diff
@@ -0,0 +1,120 @@
+// -*-c++-*--------------------------------------------------------------------
+// Copyright 2020 Bernd Pfrommer <bernd.pfrommer@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <flir_spinnaker_ros2/camera_autoexposure.h>
+
+#include <fstream>
+#include <functional>
+#include <image_transport/image_transport.hpp>
+#include <iomanip>
+#include <iostream>
+#include <math.h>
+#include <rclcpp_components/register_node_macro.hpp>
+#include <sensor_msgs/fill_image.hpp>
+#include <sensor_msgs/image_encodings.hpp>
+
+#include "logging.h"
+
+namespace flir_spinnaker_ros2
+{
+
+constexpr uint32_t CameraAutoExposure::kMinExposureUs;
+constexpr uint32_t CameraAutoExposure::kMaxExposureUs;
+constexpr float CameraAutoExposure::kMaxGain;
+constexpr uint32_t CameraAutoExposure::kExposureTransition;
+constexpr float CameraAutoExposure::kGainTransition;
+
+CameraAutoExposure::CameraAutoExposure(const rclcpp::NodeOptions& options) 
+: Node("cam_autoexposure", options)
+{
+
+  declare_parameter("camera_name", "blackfly_s");
+  get_parameter("camera_name", camera_name_);
+  declare_parameter("target_intensity", 120);
+  get_parameter("target_intensity", target_intensity_);
+  declare_parameter("intensity_update_gain", 0.5);
+  get_parameter("intensity_update_gain", intensity_update_gain_);
+  // image_sub_ = create_subscription<sensor_msgs::msg::Image>(
+  //     "/" + camera_name_ + "/image_raw", 1, std::bind(&CameraAutoExposure::imageCallback, this, std::placeholders::_1));
+  camera_meta_sub_ = create_subscription<image_meta_msgs_ros2::msg::ImageMetaData>(
+      "/" + camera_name_ + "/meta", qosDepth_, std::bind(&CameraAutoExposure::metaCallback, this, std::placeholders::_1));
+  control_pub_ = 
+    create_publisher<camera_control_msgs_ros2::msg::CameraControl>("/" + camera_name_ + "/control", qosDepth_);
+}
+
+CameraAutoExposure::~CameraAutoExposure() {
+
+}
+
+void CameraAutoExposure::metaCallback(
+  const image_meta_msgs_ros2::msg::ImageMetaData::ConstSharedPtr msg) {
+  if (msg->brightness == -1) {
+    RCLCPP_WARN(get_logger(), "Brightness is disabled, not running exposure control.");
+  }
```

Comment:
in this case, shouldn't it prevent the rest of the function's code from running?